source/log.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright 2010 - 2018 Amazon.com, Inc. or its affiliates.All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file.This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied.See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/condition_variable.h>
+#include <aws/common/linked_list.h>
+#include <aws/common/log.h>
+//#include <aws/common/memory_pool.h>
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <time.h>
+
+struct s_msg {
+    struct aws_linked_list_node node;
+    const char *tag;
+};
+
+struct s_log_ctx {
+    struct aws_allocator *alloc;
+    void *msg_pool;
+    size_t message_size;
+    struct aws_linked_list messages;
+};
+
+/* Logging globals. */
+struct s_log_ctx *s_ctx;
```

Comment:
All of these globals should probably be static.

Reply:
The pain of having to pass around the log context means this infrastructure probably won't actually get used if it's not global - or at least inherited-thread-local.